eng/scripts/splittestdependencies/tests/Generate-Dependency-Test-References-Tests.ps1

```diff
@@ -0,0 +1,102 @@
+    ) {
+        $projGroups = Split-Project-File-To-Groups -ProjectFile $projectFile -NumberOfTestsPerJob $NumOfTestProjectsPerJob `
+            -ExcludeService $excludeTargetService -ServiceDir "core" 
+        $projGroups.Length | Should -Be $length
+    }
+    # Failed cases
+    It "Zero test projects per job" -TestCases @(
+        @{ projectFile = "$PSScriptRoot/inputs/projects.txt"; NumOfTestProjectsPerJob = 0}
+        @{ projectFile = "$PSScriptRoot/inputs/empty_projects.txt"}
+    ) {
+        {Split-Project-File-To-Groups -ProjectFile $projectFile -NumberOfTestsPerJob $NumOfTestProjectsPerJob `
+            -ExcludeService $true -ServiceDir "core"} | Should -Throw -ExceptionType ([System.Management.Automation.RuntimeException])
+    }
+}
+
+Describe "Write-Test-Dependency-Group-To-Files" -Tag "UnitTest" {
+    # passed cases
+    It "Split project group array to prop xml files " -TestCases @(
+        @{ projGroupArray = @(@("1", "2"), @("3", "4")); length = 2; expectOutput = @("$PSScriptRoot/expectOutputs/1-1.xml", "$PSScriptRoot/expectOutputs/1-2.xml")}
+        @{ projGroupArray = ,@("1", "2"); length = 1; expectOutput = @("$PSScriptRoot/expectOutputs/2-1.xml")}
+    ) {
+        $outputFiles = Write-Test-Dependency-Group-To-Files -ProjectGroups $projGroupArray -MatrixOutputFolder "$PSScriptRoot/testFolder" 
+        $outputFiles.Length | Should -Be $length
+        for ($i = 0; $i -lt $length; $i++) {
+            Get-Content "$PSScriptRoot\testFolder\$($outputFiles[$i])" | Should -Be (Get-Content $expectOutput[$i])
+        }
+    }
+    # Failed cases
+    It "Empty project array" -TestCases @(
+        @{ projGroupArray = ,@(); }
+        @{ projGroupArray = @(); }
+    ) {
+        { Write-Test-Dependency-Group-To-Files -ProjectGroups $projGroupArray -MatrixOutputFolder "$PSScriptRoot/testFolder" }
+            | Should -Throw -ExceptionType ([System.Management.Automation.RuntimeException])
+    }
+}
+
+Describe "Write-Project-Files-To-Matrix" -Tag "UnitTest" {
+    # Passed cases
+    It "Write project files into matrix property" -TestCases @(
+        @{ TestFolderName = "$PSScriptRoot/testFolder1"; ProjectFiles = "1.txt", "2.txt"; MatrixJsonPath = "$PSScriptRoot/inputs/platform-matrix1.json"; expectOutput="$PSScriptRoot/expectOutputs/expect-matrix1.json" }
+        @{ TestFolderName = "$PSScriptRoot/testFolder2"; ProjectFiles = "1.txt", "2.txt", "3.txt"; MatrixJsonPath = "$PSScriptRoot/inputs/platform-matrix1.json"; expectOutput="$PSScriptRoot/expectOutputs/expect-matrix2.json" }
+    ) {
+        Write-Project-Files-To-Matrix -ProjectFiles $ProjectFiles -MatrixJsonPath $MatrixJsonPath -MatrixOutputFolder $TestFolderName -ProjectFileConfigName "OverrideFiles"
+        Get-Content "$TestFolderName\platform-matrix.json" | Should -Be (Get-Content $expectOutput)
+    }
+    #Failed cases
+    It "More include agents than project files" -TestCases @(
```

Comment:
I would say if we overrun the project file count via include, we can just loop back to the start of the project file index rather than throwing an error?

Reply:
Discussed offline. We decided to leave this as it is. Will work on optimal strategy in a follow-up.